doc/ref_arch/kubernetes/chapters/chapter06.md

```diff
@@ -10,9 +10,58 @@
+
+- Container Registry: A container registry is a repository to manage container images. The access to container registry needs to be secured in order to provide a controlled access.
+- Container Images: Stored instance of a container that holds a set of software needed to run an application. Before loading them to container registry, they need to be made secured by performing various checks like vulnerability analysis, scans etc. These should also be signed from trusted sources.
+- Containers: A lightweight and portable executable image that contains software and all of its dependencies. The containers need to be prevented from accessing the underlying OS like loading of kernel modules, mounting of directories of underlying OS etc and it must also be ensured that they don't run in priveleged mode..
+- Pods: A Pod represents a set of running containers on your cluster. Kuberenetes inherently offers pod security policies that define a set of conditions that a pod must run with in order to be accepted into the system. These policies help in ensuring the necessary checks for running the pods.
+- Kubernetes Node: A Kubernetes node in an unsecured boundary can lead to a potential threat to the running workloads. Such a node may be hardened in terms of security by disabling unused ports, prohibiting root access etc.
+- Kubernetes Master: A master node in an unsecured boundary can lead to a potential threat to the running workloads. A master may be hardened in terms of security by disabling unused ports, prohibiting root access etc.
+- Kubernetes Control Plane: The container orchestration layer that exposes the API and interfaces to define, deploy, and manage the lifecycle of containers. The communication over these APIs needs to be secured via different mechanisms like TLS encryption, API authentication via LDAP etc.
+
+<a name="6.4"></a>
+## 6.4 Isolation
+### VM vs. Container Isolation
+Sometimes container isolation is compared directly with VM based isolation, with the conclusion '*there are issues with container isolation, it is not as good as VM isolation*'. Such 1:1 comparison is not reasonable because VM and container based isolation are fundamentally different:
+- VMs: hard isolation, in the layers underlying the application SW
+- Containers: isolation by SW based mechanisms available in OS, Docker and Kubernetes. A container workload is just a set of Linux processes. It is _possible_ to configure SW based _additional isolation_ for container workloads, for example by kernel namespaces.
+
+Thus the primary isolation mechanism in Kubernetes environment should be VM or physical machine based isolation. This means: multiple container applications should not be deployed together in the same Kubernetes cluster - unless those have been planned and verified to co-exist.
```

Comment:
"Thus" is unnecessary.

"... multiple container applications should not be deployed together in the same Kubernetes cluster .."
1. even if the applications are deployed on different physical machines. Application per pod and only pods of an application on a node.
2. should this just be "applications" and not "container application"?

Reply:
"Container application" or "application" is bit ambiguous, it may refer to whole CNF or application like PostgreSQL.  I think here CNF may have been meant.  That is: in order to avoid unforeseen security consequences, CNFs can be isolated with machine level isolation as primary means.  Kernel namespaces or Kubernetes will not be able to provide same level of isolation between co-located CNFs.  

Or in other words: 
 
Lets assume that in case of VM based deployment there was some function X.  It was considered important to run X in isolation from function Y.  The level of required isolation was virtual machine level isolation i.e. sharing operating system was not seen acceptable from security perspective.

Now when deploying function X inside container, the same should be still considered.  Just containerizing X and Y  should not *automatically* remove the need for machine level isolation if that was previously considered as necessary.  Comparing kernel level isolation vs (virtual) machine level isolation is comparing apples and oranges :)

Reply:
multiple container application can be deployed in the same kubernetes cluster but seperated by namespaces. You can have multiple namespaces i.e Test, Dev, etc on the same cluster.

For container application that require more elevated privileges, we would recommend that a new container runtime be instantiated for this.

The main message should be that the security boundaries of contaniner application should be the operating system it runs on.

We can have other levels of isolation depending on the capability/requirement of the container application. For instance, we could KubeVirt/Virtlet for VM isolation or Kata container which can be used as security bundary for the application but not neccessarily a VM isolation.

Reply:
Right, multiple container applications can be deployed in the same K8s cluster. But then the risk of isolation failure is bigger because isolation mechanisms are based only on SW. So essentially it is about evaluation of such risk: what is the probability of a 'isolation failure' vulnerability to occur; and what would be the consequences.

But then there are other aspects, for example if multiple container applications are using Istio, deployment of those in the same cluster may cause clashes. Or something else which is K8s cluster wide concept like Istio.